src/plugins/ucx/ucx_backend.cpp

```diff
@@ -1817,3 +1819,33 @@ nixl_status_t nixlUcxEngine::genNotif(const std::string &remote_agent, const std
     }
     return NIXL_SUCCESS;
 }
+
+void
+nixlUcxEngine::performConnectionEstablishment(
+    const std::string &remote_agent,
+    const std::shared_ptr<nixlUcxConnection> &conn) const {
+    NIXL_DEBUG << "Establishing connection with " << remote_agent;
+
+    // Flush all endpoints to ensure connection establishment
+    // and avoid UCS_ERR_NOT_CONNECTED errors during data transfers
```

Comment:
To me it looks like trying to move workaround from user level to ucx backend instead of fixing UCP API, UCP EP should not return NOT_CONNECTED to avoid blocking on any level. Instead, the request should go on pending until completion as any other operation posted on UCP EP.